File: app/src/main/java/com/lrd/base/BaseActivity.java
package com.lrd.base;

import android.os.Bundle;
import android.widget.Toast;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;


/**
 * Created By LRD
 * on 2019/4/9  notes：
 */
public abstract class BaseActivity<T extends BasePresenter> extends AppCompatActivity implements BaseView {
	protected T mPresenter;
	@Override
	protected void onCreate(@Nullable Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(getLayoutId());
		bindView();
	}

	protected abstract int getLayoutId();

	protected abstract T getPresenter();

	protected abstract void initialize();

	@Override
	public void bindView() {
		mPresenter = getPresenter();
		mPresenter.register(this);
		initialize();
	}

	@Override
	public void unBindView() {
		mPresenter.unRegister();
	}

	@Override
	protected void onDestroy() {
		super.onDestroy();
		unBindView();
	}

	public void showToast(String text) {
		Toast.makeText(getApplicationContext(), text, Toast.LENGTH_SHORT).show();
	}
}

File: app/src/main/java/com/lrd/base/BaseActivity.kt
package com.lrd.base

import android.os.Bundle
import android.os.PersistableBundle
import android.support.v7.app.AppCompatActivity
import android.widget.Toast

/**
 * Created By LRD
 * on 2019/4/8  notes：
 */
abstract class BaseActivity<P : BasePresenter<IView>> : AppCompatActivity(),IView{
    lateinit var mPresenter : P

    override fun onCreate(savedInstanceState: Bundle?, persistentState: PersistableBundle?) {
        super.onCreate(savedInstanceState, persistentState)
        setContentView(getLayoutId())

        bindView()
    }

    abstract fun getLayoutId(): Int

    abstract fun getPresenter(): P

    abstract fun initialize()

    override fun bindView() {
        mPresenter = getPresenter()
        mPresenter.register(this)
        initialize()
    }

    override fun unBindView() {
        mPresenter.unRegister()
    }

    fun showToast(str:String){
        Toast.makeText(this,str, Toast.LENGTH_SHORT).show()
    }

    override fun onDestroy() {
        super.onDestroy()
        unBindView()
    }
}